test/IntegrationTests.t.sol

```diff
@@ -41,141 +44,421 @@ contract IntegrationTests is Test {
+        freezeAllAssetsSpell.freeze();
+    }
+
+    function test_cannotCallWithoutRoleSetup() external {
+        _vote(address(freezeAllAssetsSpell));
+
+        assertTrue(authority.hat() == address(freezeAllAssetsSpell));
+        assertTrue(
+            authority.canCall(
+                address(freezeAllAssetsSpell),
+                address(freezer),
+                freezer.freezeAllMarkets.selector
+            )
+        );
+
+        vm.expectRevert(bytes("4"));  // CALLER_NOT_RISK_OR_POOL_ADMIN
+        freezeAllAssetsSpell.freeze();
+    }
+
+}
+
+contract FreezeSingleAssetSpellTest is IntegrationTestsBase {
+
+    using SafeERC20 for IERC20;
+
+    address[] public untestedReserves;
+
+    function setUp() public override {
+        super.setUp();
         vm.prank(SPARK_PROXY);
         aclManager.addRiskAdmin(address(freezer));
+    }
 
-        assertEq(_isFrozen(WETH), false);
+    function test_freezeAssetSpell_allAssets() external {
+        address[] memory reserves = pool.getReservesList();
 
-        deal(WETH, sparkUser, 20 ether);  // Deal enough for 2 supplies
+        assertEq(reserves.length, 9);
 
-        // 1. Check user actions before freeze
-        // NOTE: For all checks, not checking pool.swapBorrowRateMode() since stable rate
-        //       isn't enabled on any reserve.
+        for (uint256 i = 0; i < reserves.length; i++) {
+            address asset = reserves[i];
 
-        vm.startPrank(sparkUser);
+            // If the asset is frozen on mainnet, skip the test
+            if (_isFrozen(asset)) {
+                untestedReserves.push(asset);
+                continue;
+            }
 
-        IERC20(WETH).safeApprove(POOL, type(uint256).max);
+            uint256 decimals = IERC20(asset).decimals();
 
-        // User can supply, borrow, repay, and withdraw
-        pool.supply(WETH, 10 ether, sparkUser, 0);
-        pool.borrow(WETH, 1 ether, 2, 0, sparkUser);
-        pool.repay(WETH, 0.5 ether, 2, sparkUser);
-        pool.withdraw(WETH, 1 ether, sparkUser);
+            uint256 supplyAmount   = 1_000 * 10 ** decimals;
+            uint256 withdrawAmount = 1 * 10 ** decimals;
+            uint256 borrowAmount   = 2 * 10 ** decimals;
+            uint256 repayAmount    = 1 * 10 ** decimals;
 
-        vm.stopPrank();
+            uint256 snapshot = vm.snapshot();
 
-        // 2. Freeze market
+            address freezeAssetSpell
+                = address(new FreezeSingleAssetSpell(address(freezer), asset));
 
-        vm.prank(randomUser);  // Demonstrate no ACL in spell
-        freezeWethSpell.freeze();
+            // Setup spell, max out supply caps so that they aren't hit during test
+            _vote(freezeAssetSpell);
 
-        assertEq(_isFrozen(WETH), true);
+            vm.startPrank(SPARK_PROXY);
+            poolConfig.setSupplyCap(asset, 68_719_476_735);  // MAX_SUPPLY_CAP
+            poolConfig.setBorrowCap(asset, 68_719_476_735);  // MAX_BORROW_CAP
```

Comment:
These can just be set to `0` to disable.

Reply:
Updated